src/full/Agda/TypeChecking/Conversion.hs

```diff
@@ -207,9 +207,13 @@ compareAs cmp a u v = do
                              | otherwise = (assign rid y vs u, assign dir x us v)
         (MetaV x us, _) -> unlessSubtyping $ assign dir x us v `orelse` fallback
         (_, MetaV y vs) -> unlessSubtyping $ assign rid y vs u `orelse` fallback
-        (Def f es, Def f' es') | f == f' ->
-          ifNotM (optFirstOrder <$> pragmaOptions) fallback $ {- else -} unlessSubtyping $ do
+        (Def f es, Def f' es') | f == f' -> do
           def <- getConstInfo f
+          foFlag <- optFirstOrder <$> pragmaOptions
+          let shortcut = foFlag || case theDef def of
+                Function{funFirstOrder = x} -> x
+                _ -> False
+          if not shortcut then fallback else unlessSubtyping $ do
```

Comment:
Please refactor this to use the short-cutting monadic conjunction `Agda.Utils.Monad.orM`.  
In particular, if `optFirstOrder` is on, you do not _also_ want to look up `getConstInfo f`.

Reply:
Ok I see now, makes sense!